Source/DataSources/GpxDataSource.js

```diff
@@ -0,0 +1,1223 @@
+            author : queryStringAttribute(node, 'author'),
+            year : queryStringValue(node, 'year', namespaces.gpx),
+            license : queryStringValue(node, 'license', namespaces.gpx)
+        };
+        if (defined(copyright.author) || defined(copyright.year) || defined(copyright.license)) {
+            return copyright;
+        } else {
+            return undefined;
+        }
+    }
+    /**
+     *  Receives a XML node and returns a boundsType object, refer to
+     * {@link http://www.topografix.com/gpx/1/1/#type_boundsType|GPX Schema}
+     */
+    function getBounds(node) {
+        var bounds = {
+            minLat : queryNumericValue(node, 'minlat'),
+            maxLat : queryNumericValue(node, 'maxlat'),
+            minLon : queryNumericValue(node, 'minlon'),
+            maxLon : queryNumericValue(node, 'maxlon')
+        };
+        if (defined(bounds.minLat) || defined(bounds.maxLat) || defined(bounds.minLon) || defined(bounds.maxLon)) {
+            return bounds;
+        } else {
+            return undefined;
+        }
+    }
+
+    var complexTypes = {
+        wpt : processWpt,
+        rte : processRte,
+        trk : processTrk
+    };
+
+    function processGpx(dataSource, node, entityCollection, sourceUri, uriResolver) {
+        var complexTypeNames = Object.keys(complexTypes);
+        var complexTypeNamesLength = complexTypeNames.length;
+
+        for (var i = 0; i < complexTypeNamesLength; i++) {
+            var typeName = complexTypeNames[i];
+            var processComplexTypeNode = complexTypes[typeName];
+
+            var childNodes = node.childNodes;
+            var length = childNodes.length;
+            for (var q = 0; q < length; q++) {
+                var child = childNodes[q];
+                if (child.localName === typeName && namespaces.gpx.indexOf(child.namespaceURI) !== -1) {
+                    processComplexTypeNode(dataSource, child, entityCollection, sourceUri, uriResolver);
+                }
+            }
+        }
+    }
+
+    function loadGpx(dataSource, gpx, sourceUri, uriResolver) {
+        var entityCollection = dataSource._entityCollection;
+
+        dataSource._promises = [];
+        entityCollection.removeAll();
+
+        var element = gpx.documentElement;
+        var version = queryStringAttribute(element, 'version');
+        var creator = queryStringAttribute(element, 'creator');
+
+        var name;
+        var metadata = processMetadata(element);
+        if(defined(metadata)){
+            name = metadata.name;
+        }
+
+        if (!defined(name) && defined(sourceUri)) {
+            name = getFilenameFromUri(sourceUri);
+        }
+
+
+        if (element.localName === 'gpx') {
+            processGpx(dataSource, element, entityCollection, sourceUri, uriResolver);
+        } else {
+            window.console.log('GPX - Unsupported node: ' + element.localName);
+        }
+
+        return when.all(dataSource._promises, function() {
```

Comment:
You can remove `_promises` and all related code, since it's not used in this reader.  Just move the context of this when call up a level and get rid of the function closure.